pkg/resource/deploy/source_eval.go

```diff
@@ -1724,6 +1737,29 @@ func (rm *resmon) resolveProvider(
 	return ""
 }
 
+func reportDiagnosticError(sink diag.Sink, urn urn.URN, t tokens.Type, name string, st *status.Status) {
+	message := st.Message()
+	for _, d := range st.Details() {
+		switch d := d.(type) {
+		case *errdetails.BadRequest:
+			message = fmt.Sprintf("%v\n\t\tBadRequest:", message)
+			for _, violation := range d.GetFieldViolations() {
+				message = fmt.Sprintf("%v\n\t\t- '%v': %v", message, violation.GetField(), violation.GetDescription())
+			}
+		case *errdetails.DebugInfo:
+			message = fmt.Sprintf("%v\n\t\t- DebugInfo: %v", message, d.GetDetail())
+			for _, entry := range d.GetStackEntries() {
+				message = fmt.Sprintf("%v\n\t\t\t%v", message, entry)
+			}
+		case *errdetails.Help:
+			for _, link := range d.GetLinks() {
+				message = fmt.Sprintf("%v\n\t\t- %v: %v", message, link.GetDescription(), link.GetUrl())
+			}
+		}
+	}
+	sink.Errorf(diag.GetResourceInvalidError(urn), t, name, message)
+}
```

Comment:
I feel like this could just be `statusToMessage`/`getStatusMessage` or similar, with the `sink.Errorf` call at the (single?) call site? A small doc comment wouldn't go amiss either IMO.

Reply:
Regardless of where we put the call, does passing an empty string to `sink.Errorf` not emit a message? Since the set of types we handle here is smaller than those that might be passed, right?